src/coroutine.rs

```diff
@@ -0,0 +1,218 @@
+    pub fn from_future<F, T, E>(future: F) -> Self
+    where
+        F: Future<Output = Result<T, E>> + Send + 'static,
+        T: IntoPy<PyObject> + Send,
+        E: Send,
+        PyErr: From<E>,
+    {
+        let wrap = async move {
+            let obj = future.await?;
+            // SAFETY: GIL is acquired when future is polled (see `Coroutine::poll`)
+            Ok(obj.into_py(unsafe { Python::assume_gil_acquired() }))
+        };
+        Self::new(Box::pin(wrap), None)
+    }
+
+    fn poll(
+        &mut self,
+        py: Python<'_>,
+        throw: Option<&PyAny>,
+    ) -> PyResult<IterNextOutput<PyObject, PyObject>> {
+        // raise if the coroutine has already been run to completion
+        let future_rs = match self.future {
+            Some(ref mut fut) => fut,
+            None => return Err(PyRuntimeError::new_err(COROUTINE_REUSED_ERROR)),
+        };
+        // pass thrown exception to throw callback if there is one, otherwise reraise it
+        if let Some(exc) = throw {
+            if let Some(cb) = &mut self.throw_callback {
+                cb(exc)
+            } else {
+                self.future.take();
+                return Err(PyErr::from_value(exc));
+            };
+        }
+        loop {
+            // create a waker from an `asyncio.Future`, or try to modify it in place
+            if let Some(waker) = self.waker.as_mut().and_then(Arc::get_mut) {
+                waker.future = asyncio_future(py)?;
+            } else {
+                self.waker = Some(Arc::new(PyFutureWaker {
+                    future: asyncio_future(py)?,
+                    thread_id: current_thread_id(),
+                }));
+            }
+            let waker = futures::task::waker(self.waker.clone().unwrap());
+            // poll the Rust future and forward its results if ready
+            if let Poll::Ready(res) = future_rs.as_mut().poll(&mut Context::from_waker(&waker)) {
+                self.future.take();
+                return Ok(IterNextOutput::Return(res?));
+            }
+            // `asyncio.Future` must be awaited; fortunately, it implements `__iter__ = __await__`
+            // and will yield itself if its result has not been set in polling above
+            let future = self.waker.as_ref().unwrap().future.as_ref(py);
+            if let Some(future) = PyIterator::from_object(future).unwrap().next() {
+                // future has not been leaked into Python for now, and Rust code can only call
+                // `set_result(None)` in `ArcWake` implementation, so it's safe to unwrap
+                return Ok(IterNextOutput::Yield(future.unwrap().into()));
+            }
+            // Python future was done, i.e. `FutureWaker::wake_by_ref` has been called in
+            // `future_rs.as_mut().poll(...)`, so continue looping
+        }
+    }
+}
+
+pub(crate) fn iter_result(result: IterNextOutput<PyObject, PyObject>) -> PyResult<PyObject> {
```

Comment:
This will our handling of iterators even more of a mess, won't it? So this is basically what we want to do with all iterators but it would apply only to coroutines? Maybe we should align all of this if we go ahead?

Reply:
(Sorry if this not clear, but this is mainly directed at @davidhewitt, not so much at @wyfo.)

Reply:
I honestly think `__next__` should allow to return `PyResult<PyObject>` in addition to `PyResult<IterNextOutput<PyObject, PyObject>>` or `PyResult<Option<PyObject>>`.
It would remove the need for this small utilitary function.
But I did want to touch too much parts of PyO3 in this first draft.

Reply:
We have a long discussion in #3190 about reworking `__next__` and `__anext__`. Really just needs someone to write up a bunch of test cases / examples so we can commit to a design. I keep meaning to do it, but am sort of trying to avoid breaking too much at the same time as #3382